### PR TITLE
Programmable Gateways 

### PR DESCRIPTION
**Pull request by sarahwooders**
**Changes:** 
* Add `Pipeline` object for queueing multiple jobs (example in `python examples/test_pipeline.py`. )
* Require `Dataplane` creation to provide a topology plan 
* Modify gateways to run on gateway programs 
* Modify planner (`planner.py`) to generate a `TopologPlan` object which can be compiled into a gateway program JSON to be provided to gateways 
* Modify `ProgressBarTransferHook` to support tracking multiple destinations 
* Support pipelining upload ID generation for multipart transfers by adding a `/api/v1/upload_id_maps` endpoint to Gateway API to send upload ID mappings to gateways along with chunk dispatching

### CODE REVIEW

File: skyplane/api/client.py [hunk 1]
```diff
@@ -60,6 +61,7 @@ def __init__(
         # set up logging
         self.log_dir.mkdir(parents=True, exist_ok=True)
         logger.open_log_file(self.log_dir / "client.log")
+        print("logging:", self.log_dir / "client.log")
```
Comment (simon-mo) [inline skyplane/api/client.py:64]: use logger

File: skyplane/api/provisioner.py [hunk 2]
```diff
@@ -155,12 +155,14 @@ def get_node(self, uuid: str) -> compute.Server:
         return self.provisioned_vms[uuid]
 
     def _provision_task(self, task: ProvisionerTask):
+        print("provision", task.region)
```
Comment (simon-mo) [inline skyplane/api/provisioner.py:158]: clenaup

File: skyplane/api/tracker.py [hunk 3]
```diff
@@ -133,91 +138,137 @@ def run(self):
...
+        with ThreadPoolExecutor(max_workers=len(dest_regions)) as executor:
+            e2e_start_time = time.time()
+            try:
+                future_list = [executor.submit(monitor_single_dst_helper, dest) for dest in dest_regions]
+                for future in as_completed(future_list):
+                    results.append(future.result())
+            except Exception as e:
+                raise e
+        e2e_end_time = time.time()
+        print("results", results)
```
Comment (simon-mo) [inline skyplane/api/tracker.py:211]: remove

File: skyplane/api/tracker.py [hunk 4]
```diff
@@ -233,31 +284,41 @@ def monitor_transfer(pd, self):
                 time.sleep(0.05)
                 continue
 
+            # TODO: have visualization for completition across all destinations
             is_complete_rec = (
-                lambda row: row["state"] == ChunkState.upload_complete
-                and row["instance"] in [s.instance for s in sinks]
-                and row["region"] in [s.region for s in sinks]
+                lambda row: row["state"] == ChunkState.complete
+                # and row["instance"] in [s.instance for s in sinks]
```
Comment (simon-mo) [inline skyplane/api/tracker.py:290]: remove this line? seems to be superceded?

File: skyplane/api/tracker.py [hunk 5]
```diff
@@ -314,9 +376,9 @@ def query_bytes_dispatched(self):
         for job_uuid in self.job_complete_chunk_ids.keys():
             bytes_total_per_job[job_uuid] = sum(
                 [
-                    cr.chunk.chunk_length_bytes
+                    cr.chunk_length_bytes
                     for cr in self.job_chunk_requests[job_uuid].values()
-                    if cr.chunk.chunk_id in self.job_complete_chunk_ids[job_uuid]
+                    # if cr.chunk_id in self.job_complete_chunk_ids[job_uuid]
```
Comment (simon-mo) [inline skyplane/api/tracker.py:381]: is this safe to remove?

File: skyplane/api/transfer_job.py [hunk 6]
```diff
@@ -429,32 +595,61 @@ def dispatch(
...
         """
-        chunker = Chunker(self.src_iface, self.dst_iface, transfer_config)
-        transfer_pair_generator = self.gen_transfer_pairs(chunker)
-        gen_transfer_list = tail_generator(transfer_pair_generator, self.transfer_list)
+        chunker = Chunker(self.src_iface, self.dst_ifaces, transfer_config)
+        transfer_pair_generator = self.gen_transfer_pairs(chunker)  # returns TransferPair objects
+        gen_transfer_list = chunker.tail_generator(transfer_pair_generator, self.transfer_list)
         chunks = chunker.chunk(gen_transfer_list)
-        chunk_requests = chunker.to_chunk_requests(chunks)
+        # chunk_requests = chunker.to_chunk_requests(chunks)
```
Comment (simon-mo) [inline skyplane/api/transfer_job.py:602]: remove

File: skyplane/api/transfer_job.py [hunk 7]
```diff
@@ -472,6 +667,7 @@ def dispatch(
 
     def finalize(self):
         """Complete the multipart upload requests"""
+        print("Finalizing multipart uploads...")
```
Comment (simon-mo) [inline skyplane/api/transfer_job.py:670]: use logger

File: skyplane/compute/aws/aws_auth.py [hunk 8]
```diff
@@ -54,12 +54,16 @@ def save_region_config(boto3, self, config: SkyplaneConfig):
...
+
+            # query regions
             describe_regions = boto3.client("ec2", region_name="us-east-1").describe_regions()
             for region in describe_regions["Regions"]:
                 if region["OptInStatus"] == "opt-in-not-required" or region["OptInStatus"] == "opted-in":
                     region_text = region["Endpoint"]
                     region_name = region_text[region_text.find(".") + 1 : region_text.find(".amazon")]
                     region_list.append(region_name)
+
+            # get VCPU limits
```
Comment (simon-mo) [inline skyplane/compute/aws/aws_auth.py:66]: this is the not the right line. should be below.

File: skyplane/compute/aws/aws_cloud_provider.py [hunk 9]
```diff
@@ -221,18 +221,25 @@ def start_instance(subnet_id: str):
...
+                        print("INVALID IAM INSTANCE PROFILE NAME")
                         logger.warning(str(e))
                     elif "InsufficientInstanceCapacity" in str(e):
                         # try another subnet
+                        print("try another subnet", region, subnets[current_subnet_id], subnets[(current_subnet_id + 1) % len(subnets)])
+                        print("all", subnets)
                         current_subnet_id = (current_subnet_id + 1) % len(subnets)
+                    else:
+                        print("UNKNOWN", e)
                     time.sleep(backoff)
```
Comment (simon-mo) [inline skyplane/compute/aws/aws_cloud_provider.py:243]: this chunk is user facing. use logger

**Review by simon-mo — APPROVED** [review]
Mostly cosmetic comments. Can we add unit test for the gateway component? It is an isolated component and it should be very easy to test standalone. This can help debugging easier as well.

**Comment by sarahwooders** [comment]
I made an issue for adding gateway components tests #806, since ideally we can get this merged asap

**Pull request merged**